Apply excludes for *.rpm files from command-line (RhBug:1262878)

j-mracek opened the pull request.

mluscon commented on dnf/base.py:
```diff
@@ -926,7 +926,9 @@ def add_remote_rpm(self, path):
+        pkg_rpm = self.sack.add_cmdline_package(path)
+        self._setup_excludes_includes()
+        return pkg_rpm
 
```
This is not a right solution e.g. it won't work for packages not presented in any enabled repository. Focus more on `_setup_excludes_includes()`.

j-mracek commented:
Exclude is applied for rpm files for Install, upgrade, reinstall, and downgrade commands. With install and exclude of rpm file the command fails; with other commands it is performed but without rpm file. Now the upgrade, reinstall, and downgrade commands also fails when strict=True and argument cannot be found.

j-mracek commented:
Personally I don't like done = True statement like in install.pu line 96, but so far I don't have a replacement for it (the result is affected by last package)

mluscon commented:
Could you please make separate commits for the fix of 1262878 and the rest? I would also appreciate a more verbose description of commits. Btw, haven't we discussed that the strict option should not influence upgrade? @jsilhan?

j-mracek commented:
@mluscon I split it into two commits

jsilhan commented:
> Btw, haven't we discussed that the strict option should not influence upgrade?

We have discussed many times that we will not do that until there's some use case and even demand.
Commit 0152105 is mostly about printing better messages, right? Can it be separated with `strict` handling? Does it even fix any of your 3 reported bugs, @j-mracek?

mluscon commented on dnf/base.py:
```diff
@@ -251,6 +251,11 @@ def fill_sack(self, load_system_repo=True, load_available_repos=True):
                     logger.warning(_("%s, disabling."), e)
         conf = self.conf
         self._sack.configure(conf.installonlypkgs, conf.installonly_limit)
+        self.rpm_packages = []
```
This should be in init prefixed with _ and preferably with some better name.

mluscon commented on dnf/base.py:
```diff
@@ -251,6 +251,11 @@ def fill_sack(self, load_system_repo=True, load_available_repos=True):
         conf = self.conf
         self._sack.configure(conf.installonlypkgs, conf.installonly_limit)
+        self.rpm_packages = []
+        if hasattr(self, 'extcmds'):
```
You are breaking the encapsulation of object here.

mluscon commented on dnf/cli/cli.py:
```diff
@@ -329,12 +329,12 @@ def distro_sync_userlist(self, userlist):
             raise dnf.exceptions.Error(msg)
 
-    def downgradePkgs(self, userlist):
+    def downgradePkgs(self, extcmds):
```
Please focus on #1262878 and fix other commands in separate commits.

mluscon commented on dnf/cli/commands/install.py:
```diff
@@ -60,15 +58,13 @@ def run(self, extcmds):
-        local_pkgs = sorted(local_pkgs, reverse=True)
-        results = map(package_install_fnc, local_pkgs)
-        done = functools.reduce(operator.or_, results, False)
+        if hasattr(self.base, 'rpm_packages'):
```
This is very error prone and bad practice too.

j-mracek commented:
@mluscon I try to refactor it. Hope this time it is better

j-mracek commented:
jenkins, test this please

j-mracek commented:
jenkins, test this please

j-mracek commented:
@mluscon @jsilhan The introduction of ``strict`` behavior for other than ``install`` commands is in commit ``cd25e2495fc6e84599dd82ec48cdbba7a2ac73c5``. Please feel free to ignore the commit if there is no any usefulness. Personally I think that ``install`` and ``reinstall`` commands should have identical behavior in aspect of ``strict`` option and arguments recognition. Similar thing could be expected from a pair of commands ``upgrade`` and ``downgrade``.

j-mracek commented:
Also this PR should pretend that some arguments are skipped silently. And this is more important that introduction of strict command. The user should always receive an info that some arguments cannot be found, handled and so on.

mluscon commented:
rebase needed

j-mracek commented:
@mluscon I rebased and refactored. There is no additional strict option for other commands. 
But I have discovered additional issue in `dnf install` when strict=False. When broken dep, packages or rpms are skipped silently and echo $? = 0.

ignatenkobrain commented:
retest this please

mluscon commented:
retest this please

mluscon commented:
In general, the exclude code looks good to me. We should consider merging _add_rpms_to_sack() and add_remote_rpm() into one api function. The rest of commits improving errors belong to other PR.

mluscon commented on dnf/cli/cli.py:
```diff
@@ -726,6 +730,7 @@ def __init__(self, base):
         self.register_command(dnf.cli.commands.HelpCommand)
         self.register_command(dnf.cli.commands.HistoryCommand)
 
+
```
This violates PEP8

jsilhan commented:
Michal is right, `add_remote_rpm` function should apply excludes as it is API function. I see 2 possibilities:
* call `_setup_excludes_includes()` lazily before any `Goal`, `Query` or `Selector` call when additional repos are loaded, conf.excludes are changed or remote packages are added
* check in `add_remote_rpm` in any conf.excludes matches the added rpm package. If so execute `sack.add_excludes`

ignatenkobrain commented:
retest this please

j-mracek commented:
depends on rpm-software-management/dnf-plugins-core#176

j-mracek commented:
Depends on rpm-software-management/dnf-plugins-extras#60

j-mracek commented:
@mluscon I refactored and created test for ci-dnf-stack. Please look on it and send me feedback. I am still not sure if this is good solution.

ignatenkobrain commented:
rebase is needed

MichaelMraka commented:
Merged.
9d92c03ac23fc036127b8a3713af679c7fb52afb
